### ISSUE TITLE
How do i build on this windows

### ISSUE THREAD

**Issue by CE1CECL** [vmware/open-vm-tools#545]
I saw mentions some mentions over at https://github.com/vmware/open-vm-tools/blob/5bf675afdc2c7d021bdac29f3bebe2d3bbb70c37/open-vm-tools/lib/include/vm_product.h#L444
however
i tried to build using msys2 and i get this 

```
CE1CECL@-1- MSYS /c/open-vm-tools/open-vm-tools
# make
Making all in lib
make[1]: Entering directory '/c/open-vm-tools/open-vm-tools/lib'
Making all in guestRpc
make[2]: Entering directory '/c/open-vm-tools/open-vm-tools/lib/guestRpc'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/c/open-vm-tools/open-vm-tools/lib/guestRpc'
Making all in auth
make[2]: Entering directory '/c/open-vm-tools/open-vm-tools/lib/auth'
/bin/sh ../../libtool  --tag=CC   --mode=compile gcc -DPACKAGE_NAME=\"open-vm-tools\" -DPACKAGE_TARNAME=\"open-vm-tools\" -DPACKAGE_VERSION=\"11.3.5\" -DPACKAGE_STRING=\"open-vm-tools\ 11.3.5\" -DPACKAGE_BUGREPORT=\"open-vm-tools-devel@lists.sourceforge.net\" -DPACKAGE_URL=\"\" -DX_DISPLAY_MISSING=1 -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_TIME_H=1 -DSTDC_HEADERS=1 -DPACKAGE=\"open-vm-tools\" -DVERSION=\"11.3.5\" -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_DLOPEN=1 -DHAVE_ECVT=1 -DHAVE_FCVT=1 -DNO_DNET=1 -DHAVE_CRYPT_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_WCHAR_H=1 -DHAVE_SYS_PARAM_H=1 -DHAVE_SYS_SYSINFO_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_VFS_H=1 -DHAVE_SYSLIMITS_H=1 -DHAVE_UNWIND_H=1 -DHAVE__BOOL=1 -DHAVE_STDBOOL_H=1 -DHAVE_STRUCT_STAT_ST_RDEV=1 -DTIME_WITH_SYS_TIME=1 -DNO_XSM=1 -DNO_XCOMPOSITE=1 -DNO_MULTIMON=1 -I.   -I/c/open-vm-tools/open-vm-tools/lib/include -I/c/open-vm-tools/open-vm-tools/lib/include  -DUSING_AUTOCONF=1 -DOPEN_VM_TOOLS -DNO_ICU -DVMX86_TOOLS  -g -O2 -Wall -Werror -Wno-pointer-sign -Wno-unused-value -fno-strict-aliasing -Wno-unknown-pragmas -Wno-uninitialized -Wno-deprecated-declarations -Wno-unused-const-variable -Wno-unused-but-set-variable -MT authPosix.lo -MD -MP -MF .deps/authPosix.Tpo -c -o authPosix.lo authPosix.c
libtool: compile:  gcc -DPACKAGE_NAME=\"open-vm-tools\" -DPACKAGE_TARNAME=\"open-vm-tools\" -DPACKAGE_VERSION=\"11.3.5\" "-DPACKAGE_STRING=\"open-vm-tools 11.3.5\"" -DPACKAGE_BUGREPORT=\"open-vm-tools-devel@lists.sourceforge.net\" -DPACKAGE_URL=\"\" -DX_DISPLAY_MISSING=1 -DHAVE_STDIO_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STRINGS_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_UNISTD_H=1 -DHAVE_SYS_TIME_H=1 -DSTDC_HEADERS=1 -DPACKAGE=\"open-vm-tools\" -DVERSION=\"11.3.5\" -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_DLOPEN=1 -DHAVE_ECVT=1 -DHAVE_FCVT=1 -DNO_DNET=1 -DHAVE_CRYPT_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_WCHAR_H=1 -DHAVE_SYS_PARAM_H=1 -DHAVE_SYS_SYSINFO_H=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_VFS_H=1 -DHAVE_SYSLIMITS_H=1 -DHAVE_UNWIND_H=1 -DHAVE__BOOL=1 -DHAVE_STDBOOL_H=1 -DHAVE_STRUCT_STAT_ST_RDEV=1 -DTIME_WITH_SYS_TIME=1 -DNO_XSM=1 -DNO_XCOMPOSITE=1 -DNO_MULTIMON=1 -I. -I/c/open-vm-tools/open-vm-tools/lib/include -I/c/open-vm-tools/open-vm-tools/lib/include -DUSING_AUTOCONF=1 -DOPEN_VM_TOOLS -DNO_ICU -DVMX86_TOOLS -g -O2 -Wall -Werror -Wno-pointer-sign -Wno-unused-value -fno-strict-aliasing -Wno-unknown-pragmas -Wno-uninitialized -Wno-deprecated-declarations -Wno-unused-const-variable -Wno-unused-but-set-variable -MT authPosix.lo -MD -MP -MF .deps/authPosix.Tpo -c authPosix.c  -DDLL_EXPORT -DPIC -o .libs/authPosix.o
In file included from authPosix.c:29:
/c/open-vm-tools/open-vm-tools/lib/include/vm_product.h:456:6: error: #error "Define a product string for this platform."
  456 | #    error "Define a product string for this platform."
      |      ^~~~~
In file included from authPosix.c:31:
/c/open-vm-tools/open-vm-tools/lib/include/posix.h: In function ‘Posix_GetHostByName’:
/c/open-vm-tools/open-vm-tools/lib/include/posix.h:248:9: error: implicit declaration of function ‘gethostbyname_r’; did you mean ‘gethostbyname2’? [-Werror=implicit-function-declaration]
  248 |    if ((gethostbyname_r(name, &he, buffer, sizeof buffer,
      |         ^~~~~~~~~~~~~~~
      |         gethostbyname2
/c/open-vm-tools/open-vm-tools/lib/include/posix.h: In function ‘Posix_FreeHostent’:
/c/open-vm-tools/open-vm-tools/lib/include/posix.h:312:14: error: passing argument 1 of ‘free’ discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
  312 |       free(he->h_name);
      |            ~~^~~~~~~~
In file included from authPosix.c:19:
/usr/include/stdlib.h:94:12: note: expected ‘void *’ but argument is of type ‘const char *’
   94 | void free (void *) _NOTHROW;
      |            ^~~~~~
authPosix.c: In function ‘Auth_AuthenticateUser’:
authPosix.c:524:28: error: implicit declaration of function ‘getspnam’; did you mean ‘getgrnam’? [-Werror=implicit-function-declaration]
  524 |          struct spwd *sp = getspnam(user);
      |                            ^~~~~~~~
      |                            getgrnam
authPosix.c:524:28: error: initialization of ‘struct spwd *’ from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
authPosix.c:527:20: error: invalid use of undefined type ‘struct spwd’
  527 |             pw = sp->sp_pwdp;
      |                    ^~
authPosix.c:532:15: error: implicit declaration of function ‘crypt’ [-Werror=implicit-function-declaration]
  532 |       namep = crypt(pass, pw);
      |               ^~~~~
authPosix.c:532:13: error: assignment to ‘const char *’ from ‘int’ makes pointer from integer without a cast [-Werror=int-conversion]
  532 |       namep = crypt(pass, pw);
      |             ^
cc1: all warnings being treated as errors
make[2]: *** [Makefile:505: authPosix.lo] Error 1
make[2]: Leaving directory '/c/open-vm-tools/open-vm-tools/lib/auth'
make[1]: *** [Makefile:501: all-recursive] Error 1
make[1]: Leaving directory '/c/open-vm-tools/open-vm-tools/lib'
make: *** [Makefile:547: all-recursive] Error 1

CE1CECL@-1- MSYS /c/open-vm-tools/open-vm-tools
#
```
how do i build open-vm-tools on windows
are there any doc's?

**Comment by lemke1458** [comment]
open-vm-tools won't build on Windows.  The OVT code base is used to build Tools for Windows, but we don't include a lot of the Windows specific code as part of the open-vm-tools project, and we use Microsoft build tools.

**Comment by CE1CECL** [comment]
> 
> 
> open-vm-tools won't build on Windows. The OVT code base is used to build Tools for Windows, but we don't include a lot of the Windows specific code as part of the open-vm-tools project, and we use Microsoft build tools.

" and we use Microsoft build tools."

wdym by "and we use Microsoft build tools."

**Comment by lemke1458** [comment]
Microsoft's compilers, linker, etc.

OVT isn't intended to be built on Windows, and its missing lots of code that is use for the VMware Tools we ship for Windows.

**Comment by CE1CECL** [comment]
Hello! I wasn't sure where to put this, but since this is a little related to windows, i decided to put this here:
I was recently looking at the now archived sourceforge of the VMware SVGA Development Kit, and Noticed something was always missing since commit one, it was mentioned in a readme, but was never provided, nor was the readme updated ever since day 1: https://sourceforge.net/p/vmware-svga/svn/1/tree/trunk/lib/README 

Look At the bottom of this: it says the following : 
```
win32 -
   A Win32 port of the VMWare SVGA reference driver. This driver runs
   in userspace, using the kernel mode interface provided by VMware's
   proprietary kernel-mode graphics driver for Windows XP.
```
Where can I find this? Looking at the root of any commit, even commit 1, over at the folders at https://sourceforge.net/p/vmware-svga/svn/1/tree/trunk/lib/ the "win32" folder is missing! Is anyone able to provide what was promised in 2009, any time soon? The person that said this was with the username of "micahjd". I had a hard time finding a good private contact email or other to ask about this now Abandoned software, since I am not a company. Thanks!

**Comment by scanlime** [comment]
Love to push against company culture to release a little bit of open source, and then get spammed and deadnamed about it for eternity 😑

Seriously tho this was at a company I left well over a decade ago. I only vaguely remember what this comment may have been referring to. It's almost certainly an internal tool that we never intended to include in the release.

Nothing in that reposititory even remotely forms a "promise", please reexamine what you are expecting. It was a tiny bit of openness carved out of a much larger mass, and none of it had to exist or ever had real support behind it.

**Comment by CE1CECL** [comment]
> Love to push against company culture to release a little bit of open source, and then get spammed and deadnamed about it for eternity 😑
> 
> Seriously tho this was at a company I left well over a decade ago. I only vaguely remember what this comment may have been referring to. It's almost certainly an internal tool that we never intended to include in the release.
> 
> Nothing in that reposititory even remotely forms a "promise", please reexamine what you are expecting. It was a tiny bit of openness carved out of a much larger mass, and none of it had to exist or ever had real support behind it.

Sorry to bug you about it, even over email. I only asked because I thought you would have it personally on you, even after all of those years, and since XP is no longer supported by Microsoft or even by VMware tools for like a couple years now, thats why I asked. I wasn't expecting a full driver, for windows 11, because no one said they would give it out. But for XP, I was just hoping for the most basic sample possible anyways. Its also fine if you cannot give it out.

**Comment by jrfonseca** [comment]
> Love to push against company culture to release a little bit of open source, and then get spammed and deadnamed about it for eternity 😑

Indeed! :smile: 

@CE1CECL , even if a Win32 port of the ancient SVGA reference code existed, it wouldn't be of any practical use to write a new XP OpenGL driver.

The best reference for OpenGL Windows drivers nowadays is Mesa, namely https://gitlab.freedesktop.org/mesa/mesa/-/tree/main/src/gallium/frontends/wgl .  Microsoft even contributed to it, and is using it in a few places.

Mesa implements Windows' OpenGL ICD (Installable Client Driver) interface -- the userspace driver interface used both on XP and later Windows.  The source for the VMware's WDDM/XPDM backends are not included though -- these were never open-sourced  -- not because there's any secret sauce, but rather because we didn't want to publish userspace <-> kernelspace interfaces, as these are subject to be changed and broken.

But there's not much secret source anyway: XPDM talks to the kernel via `ExtEscape` , WDDM talks via https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dkmthk/  .  These interfaces are essentially a _transport pipe_. which the userspace driver uses to batch commands to the kernel driver, but at the end of the day, what goes through the pipe is greatly to the userspace/kernelspace driver's discretion (particularly in XPDM, less so in WDDM.)

**Comment by CE1CECL** [comment]
> > Love to push against company culture to release a little bit of open source, and then get spammed and deadnamed about it for eternity 😑
> 
> Indeed! 😄
> 
> @CE1CECL , even if a Win32 port of the ancient SVGA reference code existed, it wouldn't be of any practical use to write a new XP OpenGL driver.
> 
> The best reference for OpenGL Windows drivers nowadays is Mesa, namely https://gitlab.freedesktop.org/mesa/mesa/-/tree/main/src/gallium/frontends/wgl . Microsoft even contributed to it, and is using it in a few places.
> 
> Mesa implements Windows' OpenGL ICD (Installable Client Driver) interface -- the userspace driver interface used both on XP and later Windows. The source for the VMware's WDDM/XPDM backends are not included though -- these were never open-sourced -- not because there's any secret sauce, but rather because we didn't want to publish userspace <-> kernelspace interfaces, as these are subject to be changed and broken.
> 
> But there's not much secret source anyway: XPDM talks to the kernel via `ExtEscape` , WDDM talks via https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dkmthk/ . These interfaces are essentially a _transport pipe_. which the userspace driver uses to batch commands to the kernel driver, but at the end of the day, what goes through the pipe is greatly to the userspace/kernelspace driver's discretion (particularly in XPDM, less so in WDDM.)

I only asked because at least for macOS, there is https://github.com/CE1CECL/vmsvga2-code which seems like VMwareGfx.kext uses, though it is rebranded, be nice to see contributions from vmware to the last updated in late 2021, macOS kext that doesn't seem to completely support Metal, and was discontinued from what I heard.
Anyways, the real reason why we want the Windows code is because there are as of today, 6 (+?) known by me, emulators/hypervisors, VMware ESX/I VMware Fusion, VMware Workstation, VMware Player, QEMU (https://github.com/CE1CECL/qemu-vmvga), and Virtualbox, which all support the VMware SVGA protocol in one way or another. I have worked so hard these last 7 months on the QEMU clone to use VMware SVGA to support 3D and compatibility down to Windows 95. It just doesn't sound right to me to say "No" to even release it ONCE, thats it, we don't care about SVGA 0/1/3/+ right now anyways, just SVGA2, since it was best supported. If you ask for versions, we ask the last version for XP from 2001, which somehow works on Windows NT4/2000 and 9x driver from 2001, and better yet?, to share the last version that supported Vista.
![Screenshot from 2023-12-20 05-10-02](https://github.com/vmware/open-vm-tools/assets/52040854/2f28a823-1dda-4d7b-85ab-735dd9dbabaf)

**Comment by jrfonseca** [comment]
@CE1CECL,

When I replied earlier, I somehow assumed that the end goal was to write a XPDM OpenGL driver.

A lot of effort went into all our guest drivers, particularly our display drivers due to the sheer size and complexity of the 3D graphics interfaces, so I understand why some third-party hypervisors opted for emulating our virtual device -- and therefore benefit from all those guest drivers and OS compatibility -- instead of replicating all that effort themselves by implementing their own drivers.

However emulation of VMware's SVGA virtual device is not something we ever endorsed or condoned.   In fact, more than once we found ourselves to suffer collateral damage of that practice, as applications deny-listed support for our virtual device, due to issues they found when running on imperfect emulations of our device on third-party hypervisors, even though those applications ran just fine on our own hypervisor/device.

I also want to make it clear that our virtual device behaves essentially the same, no matter what OSes and 3D APIs are used in the guest (or in the host OS for that matter.)  For example, it behaves the same way whether the VM has Windows XP OS running a D3D9 application, or whether it has a Linux OS running OpenGL application.  So for anybody interested in understanding how our device works, the best reference is still the three open-source resources referred in https://vmware-svga.sourceforge.net/ : the Linux kernel driver, X display driver, and Mesa 3D OpenGL driver.

**Comment by CE1CECL** [comment]
@jrfonseca 

> When I replied earlier, I somehow assumed that the end goal was to write a XPDM OpenGL driver.

Not at all, rather I am only doing the backend, not the frontend.

> A lot of effort went into all our guest drivers, particularly our display drivers due to the sheer size and complexity of the 3D graphics interfaces, so I understand why some third-party hypervisors opted for emulating our virtual device -- and therefore benefit from all those guest drivers and OS compatibility -- instead of replicating all that effort themselves by implementing their own drivers.

And I understand how much work had to be done for all of this, even for ARM64 support (SVGA III/3) which 3D doesn't work in Win32 yet still, just on x86

> However emulation of VMware's SVGA virtual device is not something we ever endorsed or condoned. In fact, more than once we found ourselves to suffer collateral damage of that practice, as applications deny-listed support for our virtual device, due to issues they found when running on imperfect emulations of our device on third-party hypervisors, even though those applications ran just fine on our own hypervisor/device.

For sure, even VBox did what I am doing right now, just using your open source backed virtual graphics device, because it has less if at all CVE's than the old VBoxVGA, and it has support down to Win95/NT4, Linux and Mac OS X, which VBox only does NT4, no 9x, and no OS X support. I also do understand the blacklist of SVGA, but I run a personal Anti-VM service which uses real hardware on QEMU only except for graphics, and for those that don't make sence, I change QEMU branding to be "CECL HardDrive" for example. If I could, I would use a ATI card for emulation for example, instead of the SVGA, like I do use Intel 8574L instead of vmxnet3, because it comes with real hardware, the problem is graphics suck in all VM programs. VMware today still does the best at it, because it has support for Aero, and D3D, and has again, less CVEs. in the case of the NIC, on QEMU, it performs the same as on real hardware, so thats only why i use what I have. Also NICs have better Windows backwards compat, than Graphics, which VMware SVGA is the ONLY device that supports windows 11, and windows 95, even win2k cards dont work on Windows 8. Its just how earth works to this day.

> I also want to make it clear that our virtual device behaves essentially the same, no matter what OSes and 3D APIs are used in the guest (or in the host OS for that matter.) For example, it behaves the same way whether the VM has Windows XP OS running a D3D9 application, or whether it has a Linux OS running OpenGL application. So for anybody interested in understanding how our device works, the best reference is still the three open-source resources referred in https://vmware-svga.sourceforge.net/ : the Linux kernel driver, X display driver, and Mesa 3D OpenGL driver.

Why did you delete the Git portion of the website, which was more up to date than SVN, excluding the page linked above? Some people still (for some reason) using those as ref, until it was deleted by you. Its not a big deal, but it was a noticeable thing.

**Comment by CE1CECL** [comment]
@jrfonseca Also in terms of legal, what is the license on the SVGA bins for windows, like am I able to mod them and redist my mods for something such as anti-vm? Thats another thing I was wondering about.

For anyone else reading, the old mirrors for those deleted files I have: 
https://github.com/CE1CECL/vmware-svga-20090413
https://github.com/CE1CECL/vmware-svga-git
https://github.com/CE1CECL/vmware-svga-web
https://github.com/CE1CECL/vmware-svga-svn

**Comment by jrfonseca** [comment]
> Why did you delete the Git portion of the website, which was more up to date than SVN, excluding the page linked above? Some people still (for some reason) using those as ref, until it was deleted by you. Its not a big deal, but it was a noticeable thing.

The code on https://sourceforge.net/projects/vmware-svga/ existed mostly for historical purposes. But it wasnow  decided by the team through consensus to delete all traces as it is causing more harm than good, because it is a misleading reference both for how to use (or even mimic) the virtual device, as the interfaces used are all deprecated (particular the way DMAs are done), yet folks keep stumbling into this code and look at it (perhaps because it more self contained?), and make wrong inferences from it.

The only reason I didn't remove all SVN altogether too is because the HTML that seen on https://vmware-svga.sourceforge.net/ is also hosted  on SVN and I couldn't be bothered to rewrite the SVN history to prune it.

> Also in terms of legal, what is the license on the SVGA bins for windows, like am I able to mod them and redist my mods for something such as anti-vm?

Please read the EULA(s) provided and/or displayed while installing the drivers in question.  I don't know them from the top of my head and even if I did, I'm not qualified to provide legal advice.

> For anyone else reading, the old mirrors for those deleted files I have

I don't understand why you're holding on to this old reference code as to dear life.  I admit Linux/Mesa code are larger and more complex code bases, but these are proper production quality code, which uses both the new interfaces and old interfaces (for backwards compatibility.)  So it is a must more trustworthy reference code for everything.

And you're certainly doing a disservice to us by hosting a backup.

**Comment by jrfonseca** [comment]
All I can say is that you need to honour whatever legal disclaimers (including copyright notices) that are shown, and whatever T&Cs you agree while downloading, installing, and using our products (be it Tools, Workstation, or Player).  It's not my place to provide legal advice or interpretation.

I've answered the questions the best I could.   We'll close this shortly.

**Comment by johnwvmw** [comment]
Per the previous comment, closing this issue.

**Comment by CE1CECL** [comment]
> Per the previous comment, closing this issue.

Alrighty then :)

**Comment by CE1CECL** [comment]
I caught a bug that seems to affect your VMwareGfx.kext that is inside VMware tools for darwin. 
However, I don't know if it will get fixed due to the fact the kext was discontinued by you guys a couple years ago. The bug relates to setting a custom resolution to 1366 x 768, which also did exist in the open source version on sourceforge: VMsvga2, see here for the patch I did to fix it on my rebranded VMsvga2, called CEsvga2: https://github.com/CE1CECL/cesvga2/commit/dc2e4ee7592255039030a6f870c653ca1721e907
This bug was confirmed by me on Virtualbox's SVGA, QEMU-vmvga (https://github.com/CE1CECL/qemu-vmvga) and confirmed by someone else I asked to confirm on VMware workstation (patched for OS X to work) and VMware Fusion.
The 32U in the patch is HARDCODED because the kext hardcodes 32bpp from what I am seeing so far.
(I published this issue here since I don't have a @vmware.com email for SVGA for darwin to chat privately about)
